components/fxa-client/android/src/main/java/mozilla/appservices/fxa/FirefoxAccount.kt

```diff
@@ -0,0 +1,153 @@
+ * events.  See also:
+ *   - StateManager: manages the account state machine.  Defined in `State.kt` and available from
+ *     the `stateManager()` method.
+ *   - DeviceManager: manages the list of connected devices to an FxA account.  Defined in
+ *     `Devices.kt` and available from the `deviceManager()` method.
+ */
+abstract class FirefoxAccount(
+    private val config: FxaConfig,
+    private val deviceConfig: DeviceConfig,
+    private val storageHandler: StorageHandler,
+    private val applicationScopes: List<Scope> = listOf(),
+) {
+    /**
+     * Tries to fetch an access token for the given scope.
+     *
+     * @param singleScope Single OAuth scope (no spaces) for which the client wants access
+     * @return [AccessTokenInfo] that stores the token, along with its scope, key and
+     *                           expiration timestamp (in seconds) since epoch when complete
+     */
+    abstract suspend fun getAccessToken(singleScope: String): AccessTokenInfo?
+
+    /**
+     * Fetches the profile object for the current client either from the existing cached state
+     * or from the server (requires the client to have access to the profile scope).
+     *
+     * @param ignoreCache Fetch the profile information directly from the server
+     * @return Profile (optional, if successfully retrieved) representing the user's basic profile info
+     */
+    abstract suspend fun getProfile(ignoreCache: Boolean = false): Profile?
+
+    /**
+     * Fetches the token server endpoint, for authentication using the SAML bearer flow.
+     *
+     * @return Token server endpoint URL string, `null` if it couldn't be obtained.
+     */
+    abstract suspend fun getTokenServerEndpointURL(): String?
+
+    /**
+     * Get the pairing URL to navigate to on the Authority side (typically a computer).
+     *
+     * @return The URL to show the pairing user
+     */
+    abstract fun getPairingAuthorityURL(): String
+
+    /**
+     * Returns current FxA Device ID for an authenticated account.
+     *
+     * @return Current device's FxA ID, if available. `null` otherwise.
+     */
+    abstract fun getCurrentDeviceId(): String?
+
+    /**
+     * Returns session token for an authenticated account.
+     *
+     * @return Current account's session token, if available. `null` otherwise.
+     */
+    abstract fun getSessionToken(): String?
+
+    /**
+     * Indicates if account needs to be re-authenticated via [beginAuthentication].
+     * Most common reason for an account to need re-authentication is a password change.
+     *
+     * TODO this may return a false-positive, if we're currently going through a recovery flow.
+     * Prefer to be notified of auth problems via [AccountObserver], which is reliable.
+     *
+     * @return A boolean flag indicating if account needs to be re-authenticated.
+     */
+    abstract fun accountNeedsReauth(): Boolean
```

Comment:
This seems a little wrong to me - the [existing function in a-c](https://github.com/mozilla-mobile/firefox-android/blob/d6b6f3c63cae8203c989775d9d24833a6c56dcb7/android-components/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/manager/FxaAccountManager.kt#L312) checks some other internal state enum, which in that case has a `AuthenticationProblem` - but our state enum has no equivalent state - which IMO it probably should.

Reply:
`AccountState.Disconnected` has the `fromAuthIssues` flag, which can represent that.  I tried to lower the number of top-level states, since I think a lot of code isn't only going to care if the account is connected vs disconnected.  I'm not sure if that's a good idea or not though, what do you think?

However we decide to handle the state enum, maybe this method should be removed since `getState()` can tell you the same information.